pkg/agent/controller/networkpolicy/multicast.go

```diff
@@ -0,0 +1,337 @@
+	groupAddress net.IP, direction crdv1beta.Direction) (*v1alpha1.RuleAction, apitypes.UID, *crdv1beta.NetworkPolicyType, string, error) {
+
+	var ruleTypePtr *crdv1beta.NetworkPolicyType
+	action, uuid, ruleName := v1alpha1.RuleActionAllow, apitypes.UID(""), ""
+	if iface == nil {
+		// Check the if iface exists
+		klog.ErrorS(fmt.Errorf("iface should not be empty"), "")
+		return nil, "", nil, "", fmt.Errorf("iface should not be empty")
+	}
+	ns, podname := iface.PodNamespace, iface.PodName
+	obj, exists, _ := c.mcastItemRuleCache.GetByKey(groupAddress.String())
+	if !exists {
+		klog.V(2).InfoS("Rule for group does not exist", "group", groupAddress.String())
+		action = v1alpha1.RuleActionAllow
+		return &action, "", nil, "", nil
+	}
+	item := obj.(*mcastItem)
+	var matchedRule *CompletedRule
+	member := &crdv1beta.GroupMember{
+		Pod: &crdv1beta.PodReference{
+			Name:      podname,
+			Namespace: ns,
+		},
+	}
+	for ruleID := range item.ruleIDs {
+		// Iterate all rules relate to this multicast group address
+		r, _, _ := c.ruleCache.GetCompletedRule(ruleID)
+		// Find the rule with highest priority
+		if direction == r.Direction && r.TargetMembers.Has(member) == true {
+			if (matchedRule == nil) || *(item.ruleIDs[ruleID]) > *(item.ruleIDs[matchedRule.ID]) {
+				matchedRule = r
+			}
+		}
+	}
+	if matchedRule != nil {
+		ruleTypePtr = new(crdv1beta.NetworkPolicyType)
+		action, uuid, *ruleTypePtr, ruleName = *matchedRule.Action, matchedRule.PolicyUID, matchedRule.SourceRef.Type, matchedRule.Name
+	} else {
+		action, uuid, ruleName = v1alpha1.RuleActionAllow, "", ""
+	}
+	klog.V(2).InfoS("Call validation:", "action", action, "uuid", uuid, "ruleName", ruleName, "ruleType", ruleTypePtr)
+	return &action, uuid, ruleTypePtr, ruleName, nil
+}
+
+func (c *multicastController) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *multicastController) processNextWorkItem() bool {
+	obj, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(obj)
+
+	// We expect igmpRule to come off the workqueue.
+	if item, ok := obj.(*mcastIGMPRule); !ok {
+		// As the item in the workqueue is actually invalid, we call Forget here else we'd
+		// go into a loop of attempting to process a work item that is invalid.
+		// This should not happen.
+		c.queue.Forget(obj)
+		klog.Errorf("Expected mcastIGMPRule in work queue but got %#v", obj)
+		return true
+	} else if err := c.syncIGMPRule(item); err == nil {
+		// If no error occurs we Forget this item, it does not get queued again until
+		// another change happens.
+		c.queue.Forget(item)
+	} else {
+		// Put the item back on the workqueue to handle any transient errors.
+		c.queue.AddRateLimited(item)
+		klog.ErrorS(err, "error syncing multicast rule", "item", item)
+	}
+	return true
+}
+
+func (c *multicastController) syncIGMPRule(igmpRule *mcastIGMPRule) error {
+	c.ruleIDGroupMapMutex.Lock()
+	defer c.ruleIDGroupMapMutex.Unlock()
```

Comment:
There is only singler worker, and the lock is only acquired here, so seems useless.
I guess what you should prevent is the race condition between the worker and the `Validate` method that reads the cache, but you don't use any lock there.

Reply:
It has lock inside, but whether another lock is required depends on whether a transaction is required when you consume this cache and other data together, for example, currently you have two maps which have have associated data, but I think one of them is not required.